### PR TITLE
[24] [Integrate] Forgot Password screen

### PR DESCRIPTION
**Pull request by markgravity**
https://github.com/markgravity/flutter-survey/issues/24

## What happened

Integrate Forgot Password screen 

 
## Proof Of Work

https://user-images.githubusercontent.com/17875522/117240820-e6100a80-ae5b-11eb-944a-bc6e2668fc27.mp4

### CODE REVIEW

File: lib/modules/forgot_password/forgot_password_interactor.dart [hunk 1]
```diff
@@ -1,8 +1,28 @@
...
+    extends Interactor<ForgotPasswordInteractorDelegate> {
+  void resetPassword({required String email});
+}
 
-abstract class ForgotPasswordInteractorDelegate {}
+abstract class ForgotPasswordInteractorDelegate {
+  BehaviorSubject<void> get passwordDidReset;
 
-class ForgotPasswordInteractorImpl extends ForgotPasswordInteractor {}
+  BehaviorSubject<Object> get passwordDidFailToReset;
```
Comment (manh-t) [inline lib/modules/forgot_password/forgot_password_interactor.dart:11]: avoid using `Object` here pls 🙋

Comment (markgravity) [reply lib/modules/forgot_password/forgot_password_interactor.dart]: I updated it! 🙇

File: lib/modules/forgot_password/forgot_password_presenter.dart [hunk 2]
```diff
@@ -5,6 +5,14 @@ abstract class ForgotPasswordPresenter extends Presenter<ForgotPasswordView,
 
 class ForgotPasswordPresenterImpl extends ForgotPasswordPresenter
     implements ForgotPasswordViewDelegate, ForgotPasswordInteractorDelegate {
+  ForgotPasswordPresenterImpl() {
+    emailTextFieldDidChange.listen(_emailTextFieldDidChange);
+    resetButtonDidTap.listen(_resetButtonDidTap);
+
+    passwordDidReset.voidListen(_passwordDidReset);
+    passwordDidFailToReset.listen(_passwordDidFailToReset);
```
Comment (manh-t) [inline lib/modules/forgot_password/forgot_password_presenter.dart:13]: ah here [reference](https://github.com/markgravity/flutter-survey/pull/56#discussion_r638573902)

**Review by manh-t — APPROVED** [review]
LGTM 🚀

File: lib/modules/forgot_password/forgot_password_interactor.dart [hunk 3]
```diff
@@ -1,8 +1,28 @@
...
-class ForgotPasswordInteractorImpl extends ForgotPasswordInteractor {}
+  BehaviorSubject<Exception> get passwordDidFailToReset;
+}
+
+class ForgotPasswordInteractorImpl extends ForgotPasswordInteractor {
+  final AuthRepository _authRepository = locator.get();
+
+  @override
+  void resetPassword({required String email}) {
+    Future.microtask(() async {
```
Comment (sleepylee) [inline lib/modules/forgot_password/forgot_password_interactor.dart:19]: I think we mentioned about this microtask()

File: lib/modules/forgot_password/forgot_password_presenter.dart [hunk 4]
```diff
@@ -13,4 +21,30 @@ class ForgotPasswordPresenterImpl extends ForgotPasswordPresenter
 
   @override
   final alertDialogDidClose = BehaviorSubject<void>();
+
+  @override
+  final passwordDidReset = BehaviorSubject<void>();
+
+  @override
+  final passwordDidFailToReset = BehaviorSubject<Exception>();
```
Comment (sleepylee) [inline lib/modules/forgot_password/forgot_password_presenter.dart:29]: as I mentioned in the prev PR, the point of enableForgotPassword button should be in this PR (Integrate) task, and imo, it should stay somewhere here no?

File: lib/modules/forgot_password/forgot_password_module.dart [hunk 5]
```diff
@@ -13,6 +13,8 @@ import 'package:survey/core/viper/module.dart';
 import 'package:survey/gen/assets.gen.dart';
 import 'package:survey/modules/screen.dart';
 import 'package:flutter_gen/gen_l10n/app_localizations.dart';
+import 'package:survey/repositories/auth_repository.dart';
+import 'package:survey/services/locator/locator_service.dart';
```
Comment (sleepylee) [inline lib/modules/forgot_password/forgot_password_module.dart:17]: wondering why are we just started to import this in this PR but there is no usage? 🤔

File: lib/modules/forgot_password/forgot_password_presenter.dart (hunk 4 above)
Comment (markgravity) [reply lib/modules/forgot_password/forgot_password_presenter.dart:29]: @sleepylee sorry, don't get your point here 😬

File: lib/modules/forgot_password/forgot_password_module.dart (hunk 5 above)
Comment (markgravity) [reply lib/modules/forgot_password/forgot_password_module.dart:17]: @sleepylee it's used in [here](https://github.com/markgravity/flutter-survey/blob/a36d21e16a0a73cba713e5ae6162924288902e67/lib/modules/forgot_password/forgot_password_interactor.dart)

File: lib/modules/forgot_password/forgot_password_presenter.dart (hunk 4 above)
Comment (sleepylee) [reply lib/modules/forgot_password/forgot_password_presenter.dart:29]: before, we mentioned control extending the state management in View, and also, you do the state binding in a [UI] PRs, making it very big to follow.

Comment (markgravity) [reply lib/modules/forgot_password/forgot_password_presenter.dart:29]: @sleepylee ah yea, I realized this after did it, so I changed it in my next implements

**Review by sleepylee — APPROVED** [review]

**Pull request merged**